### ISSUE TITLE
Feature request : t-sne

### ISSUE THREAD

**Issue by kojix2** [yoshoku/rumale#7]
New algorithm request
I want t-sne rather than neural networks. 

* I want to visualize the open-access cancer mRNA data from the TCGA (The Cancer Genome Atlas).

**Comment by yoshoku** [comment]
That's a good suggestion, I also want to implement t-SNE.
However, t-SNE implemented with a simple gradient method requires a lot of execution time.
Thus, I survery and consider another optimization algorithm for t-SNE.
I am planning to implement t-SNE, so please wait.

**Comment by mrkn** [comment]
How do you think about UMAP for the alternative to t-SNE?

**Comment by kojix2** [comment]
Don’t over do it 🐰

**Comment by yoshoku** [comment]
I think that the important thing in the UMAP algorithm is the initialization of low dimensional space by spectral embedding. Since the spectral embedding is realized by eigenvalue decomposition of Laplacian matrix, to implement that, it is necessary to consider the introduction of Numo::Linalg. Installing Numo::Linalg is still difficult for users unfamiliar with BLAS and LAPACK libraries. Thus, for now, I am giving priority to implement algorithms that do not require linear algebra techniques, such as eigenvalue decomposition and inverse matrix calculation.
I hope to appear an all-in-one package for data science and machine learning in Ruby like Anaconda in Python.
By the way, I am implementing t-SNE with fixed-point optimization that converges faster than gradient optimization 💪

**Comment by mrkn** [comment]
You may use [rubydata/minimal-notebook docker image](https://hub.docker.com/r/rubydata/minimal-notebook) for a development environment with Numo::Linalg.

**Comment by yoshoku** [comment]
I have released Rumale 0.10.0 including t-SNE a while ago. The t-SNE implemented in Rumale obtains low-dimensional representation using fixed-point optimization. Therefore, unlike scikit-learn, that does not have learning rate and moment term as hyper-parameters. The usage is as follows:

```ruby
tsne = Rumale::Manifold::TSNE.new(
  perplexity: 40.0, max_iter: 1000, verbose: true, random_seed: 1)
low_samples = tsne.fit_transform(samples)
```

If you find a bug please report me. Thank you.

**Comment by kojix2** [comment]
Thanks a lot!

![mnist2](https://user-images.githubusercontent.com/5798442/57964104-7781cb80-796a-11e9-8115-fe975f291f56.gif)
